### ISSUE TITLE
building libpod using podman fails, if image already present.

### ISSUE THREAD

**Issue by kunalkushwaha** [containers/libpod#1973]
**Is this a BUG REPORT or FEATURE REQUEST?**:

[//]: **kind bug**

**Description**
If `libpod` is build using `libpod`, it fails if image is build once with following error

If image cache exists, error happens while resolving the symlinks.

```
error checking if cached image exists from a previous build: error checking if copied files match: error resolving symlinks and comparing modified times: "/home/kunal/go/src/github.com/containers/libpod": panic: runtime error: slice bounds out of range
```

**Steps to reproduce the issue:**

1. sudo make dbuild

2. after completion of build, make some changes in code.

3. sudo make dbuild.

**Describe the results you received:**

```
sudo make dbuild
.
.
.
.
STEP 9: FROM 338ec34b19a9b4487720de6f7e16bd6e92083aa6ae971cfd99cd4bb95bd2eb17
STEP 10: RUN echo 'deb http://httpredir.debian.org/debian jessie-backports main' > /etc/apt/sources.list.d/backports.list
--> Using cache 0b059236b43fe02da36f450c41ae633027b1092c0bb1885d4beaec369c036834
STEP 11: FROM 0b059236b43fe02da36f450c41ae633027b1092c0bb1885d4beaec369c036834
STEP 12: RUN apt-get update && apt-get install -y     apparmor     autoconf     automake     bison     build-essential     curl     e2fslibs-dev     gawk     gettext     go-md2man     iptables     pkg-config     libaio-dev     libcap-dev     libfuse-dev     libnet-dev     libnl-3-dev     libostree-dev     libprotobuf-dev     libprotobuf-c0-dev     libtool     libudev-dev     protobuf-c-compiler     protobuf-compiler     python-minimal     libglib2.0-dev     libapparmor-dev     btrfs-tools     libdevmapper1.02.1     libdevmapper-dev     libgpgme11-dev     liblzma-dev     netcat     socat     python3-pip     python3-dateutil     python3-setuptools     python3-psutil     python3-pytoml     lsof     xz-utils     --no-install-recommends     && apt-get clean
--> Using cache c97667b08dbd449bd40ca7992b4836c2b18d2d34ab5ad42cac7d6bfe18071a95
STEP 13: FROM c97667b08dbd449bd40ca7992b4836c2b18d2d34ab5ad42cac7d6bfe18071a95
STEP 14: ADD . /go/src/github.com/containers/libpod
error checking if cached image exists from a previous build: error checking if copied files match: error resolving symlinks and comparing modified times: "/home/kunal/go/src/github.com/containers/libpod": panic: runtime error: slice bounds out of range

goroutine 1 [running]:
github.com/containers/libpod/vendor/github.com/containers/buildah/imagebuildah.modTimeIsGreater(0x7ffe3556edef, 0x2f, 0x7ffe3556ee1f, 0x2f, 0x7ffe3556ee4f, 0x1e, 0xc4202bf698, 0x41cbec, 0x415a14)
        /go/src/github.com/containers/libpod/vendor/github.com/containers/buildah/imagebuildah/chroot_symlink.go:142 +0x3a2
github.com/containers/libpod/vendor/github.com/containers/buildah/imagebuildah.resolveSymlinkTimeModified()
        /go/src/github.com/containers/libpod/vendor/github.com/containers/buildah/imagebuildah/chroot_symlink.go:98 +0x211
github.com/containers/libpod/vendor/github.com/containers/storage/pkg/reexec.Init(0x1133a80)
        /go/src/github.com/containers/libpod/vendor/github.com/containers/storage/pkg/reexec/reexec.go:26 +0x77
main.main()
        /go/src/github.com/containers/libpod/cmd/podman/main.go:51 +0x71
: exit status 2
Makefile:150: recipe for target 'libpodimage' failed
make: *** [libpodimage] Error 125
```

**Describe the results you expected:**
Build should not fail.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
$ sudo podman version
Version:       0.12.2-dev
Go Version:    go1.10.5
Git Commit:    "1c805191ba732aba770cab421cbfdf057dad6f3a-dirty"
Built:         Tue Dec 11 16:51:48 2018
OS/Arch:       linux/amd64
```

**Output of `podman info`:**
```
$ sudo podman info
host:
  BuildahVersion: 1.6-dev
  Conmon:
    package: Unknown
    path: /usr/libexec/podman/conmon
    version: 'conmon version 1.12.0-dev, commit: 6515d2577507771758b922cedf4874fa1811bce7'
  Distribution:
    distribution: elementary
    version: "5.0"
  MemFree: 1193598976
  MemTotal: 16529944576
  OCIRuntime:
    package: Unknown
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc5+dev
      commit: f3ce8221ea760f51b5403a9892a69ace43845c2c
      spec: 1.0.1-dev
  SwapFree: 2147217408
  SwapTotal: 2147479552
  arch: amd64
  cpus: 8
  hostname: kunal-HP-dev
  kernel: 4.15.0-42-generic
  os: linux
  rootless: false
  uptime: 27h 13m 42.9s (Approximately 1.12 days)
insecure registries:
  registries: []
registries:
  registries:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ContainerStore:
    number: 16
  GraphDriverName: overlay
  GraphOptions: null
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
  ImageStore:
    number: 74
  RunRoot: /var/run/containers/storage
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by mheon** [comment]
This seems like it could be a Buildah error - @TomSweeneyRedHat agree?

**Comment by TomSweeneyRedHat** [comment]
@mheon, maybe.  I'm a bit confused to what exactly is happening though.  @kunalkushwaha is this something you can reproduce outside of the makefile?  I.e. do you have a Dockerfile and a few podman/buildah commands that you use to reproduce this?

**Comment by mheon** [comment]
@TomSweeneyRedHat This looks like `make dbuild` exploding, which basically amounts to a `podman build` plus a `podman run` - it looks like the `podman build` part is what's blowing up, which led me to suspect Buildah. The exact line in question is https://github.com/containers/libpod/blob/master/Makefile#L149-L150

**Comment by rhatdan** [comment]
Could someone attempt this with the nodev line removed from /etc/containers/storage.conf.

We are seeing some weird interaction between builds being done inside of containers, if the container Overlay is mounted nodev.  (CRI-O Is seeing it, and Ben was able to reproduce with buildah.)  I will be investigating what is going on next week when back from Kubecon.

**Comment by kunalkushwaha** [comment]
Just wanted to update that in my setup,  my `/etc/containers/storage.conf `is blank i.e. it is using defaults from code and my `podman info ` output suggest dont have graphOption for `nodev`.

I think the problem lies in cache time which is set .. could be `time variable` have some corrupt value.

```
store:
  ContainerStore:
    number: 16
  GraphDriverName: overlay
  GraphOptions: null
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
  ImageStore:
    number: 74
  RunRoot: /var/run/containers/storage
```

**Comment by baude** [comment]
where are we on this?

**Comment by mheon** [comment]
I think it's definitely a Buildah bug, though I don't think we have a consistent reproducer yet

**Comment by vrothberg** [comment]
@cgwalters dropped something at https://github.com/containers/libpod/issues/2012#issuecomment-447956685 that might help here.

**Comment by baude** [comment]
@TomSweeneyRedHat are you looking at this?

**Comment by TomSweeneyRedHat** [comment]
@baude I'd looked into it before PTO and wasn't able to reproduce it in my environment.

**Comment by sjenning** [comment]
I am able to recreate.

I put a debug message in `modTimeIsGreater` and got found out that `path` and `rootdir` are the same, which makes `path[len(rootdir)+1:]` invalid.

Both are `/home/sjennings/projects/go/src/github.com/openshift/machine-config-operator` in my case.

**Comment by mheon** [comment]
Thanks for the reproducer!

**Comment by sjenning** [comment]
This code area went it here
https://github.com/containers/buildah/commit/0ce1071650c5ab8b7f5e696855e1e05198e2f1f4

fyi @umohnani8

**Comment by sjenning** [comment]
This patch fixes it for me
```diff
diff --git a/vendor/github.com/containers/buildah/imagebuildah/chroot_symlink.go b/vendor/github.com/containers/buildah/imagebuildah/chroot_symlink.go
index edb5837d..3063d3bf 100644
--- a/vendor/github.com/containers/buildah/imagebuildah/chroot_symlink.go
+++ b/vendor/github.com/containers/buildah/imagebuildah/chroot_symlink.go
@@ -136,6 +136,10 @@ func modTimeIsGreater(rootdir, path string, historyTime string) (bool, error) {
        if err != nil {
                return false, errors.Wrapf(err, "error converting string to time.Time %q", historyTime)
        }
+
+       if rootdir == path {
+               return false, nil
+       }
        // Walk the file tree and check the time stamps.
        // Since we are chroot in rootdir, only want the path of the actual filename, i.e path - rootdir.
        // +1 to account for the extra "/" (e.g rootdir=/home/user/mydir, path=/home/user/mydir/myfile.json)
```
But this seems like a hack and that the code path doesn't expect this in general.

Probably need @umohnani8 or someone on the buildah team to look at it.

**Comment by TomSweeneyRedHat** [comment]
I'm not able to reproduce using upstream.  Did someone sneak a fix in or  is my testing not right?  Then again there's not a lot of output, I'm not sure if this a "success" or a quiet failure.

```
# pwd
/root/tsweeney/workspaces/openshift/src/github.com/openshift/machine-config-operator

# make machine-config-operator
WHAT=machine-config-operator /root/tsweeney/workspaces/openshift/src/github.com/openshift/machine-config-operator/hack/build-go.sh
Using version from git...
Building github.com/openshift/machine-config-operator/cmd/machine-config-operator (v3.11.0-371-g32b8e0ba)

# make machine-config-server
WHAT=machine-config-server /root/tsweeney/workspaces/openshift/src/github.com/openshift/machine-config-operator/hack/build-go.sh
Using version from git...
Building github.com/openshift/machine-config-operator/cmd/machine-config-server (v3.11.0-371-g32b8e0ba)

# make machine-config-daemon
WHAT=machine-config-daemon /root/tsweeney/workspaces/openshift/src/github.com/openshift/machine-config-operator/hack/build-go.sh
Using version from git...
Building github.com/openshift/machine-config-operator/cmd/machine-config-daemon (v3.11.0-371-g32b8e0ba)

# make machine-config-controller
WHAT=machine-config-controller /root/tsweeney/workspaces/openshift/src/github.com/openshift/machine-config-operator/hack/build-go.sh
Using version from git...
Building github.com/openshift/machine-config-operator/cmd/machine-config-controller (v3.11.0-371-g32b8e0ba)

# podman version
Version:       0.12.2-dev
Go Version:    go1.10.5
Git Commit:    "fa998f224fb7aa15d85b09dfdc4491dc09c9e390"
Built:         Thu Dec 20 14:24:53 2018
OS/Arch:       linux/amd64

```

**Comment by rhatdan** [comment]
@sjenning Could you open a PR for imagebuilder with this fix, and then we can get it vendored into buildah?

**Comment by sjenning** [comment]
@rhatdan you mean open a PR against buildah right? the changed file is in the buildah repo.
https://github.com/containers/buildah/blob/master/imagebuildah/chroot_symlink.go

**Comment by rhatdan** [comment]
Yes sorry.

**Comment by vorburger** [comment]
I'm also hitting (I think) this one.. slightly different way than some of what is described above: 

So I never yet used `podman run` and am just starting and trying to `podman build` (actually `docker build`, via cool `dnf install podman-docker`; but I doubt that's relevant here).. that seemed to take forever and got stuck with high CPU spinning (do you want a separate issue about that?), so I Ctrl-C it - and I now, reproducibly, hit this:

```
$ cd /home/vorburger/dev/Go/src/git.opendaylight.org/gerrit/p/coe.git/odlCNIPlugin/odlovs-cni
$ podman build -t odlovs-cni . 
STEP 1: FROM golang:1 AS builder
STEP 2: RUN curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
--> Using cache 668c4e1c8676704015d5f76130eed11712a7b8a1337aab07928ba732d8cdcbac
STEP 3: FROM 668c4e1c8676704015d5f76130eed11712a7b8a1337aab07928ba732d8cdcbac AS builder
STEP 4: WORKDIR /go/src/git.opendaylight.org/gerrit/p/coe.git/odlCNIPlugin/odlovs-cni
--> Using cache 6cd9d3574dd90df62e3fea60d65a09f05b3431833d6cb30c874ea78feb590451
STEP 5: FROM 6cd9d3574dd90df62e3fea60d65a09f05b3431833d6cb30c874ea78feb590451 AS builder
STEP 6: COPY . .
error checking if cached image exists from a previous build: error checking if copied files match: error resolving symlinks and comparing modified times: "/home/vorburger/dev/Go/src/git.opendaylight.org/gerrit/p/coe.git/odlCNIPlugin/odlovs-cni": panic: runtime error: slice bounds out of range

goroutine 1 [running]:
github.com/containers/libpod/vendor/github.com/containers/buildah/imagebuildah.modTimeIsGreater(0x7ffe67cf3740, 0x58, 0x7ffe67cf3799, 0x58, 0x7ffe67cf37f2, 0x1d, 0xc000299718, 0x4357c1, 0x41d511)
	/builddir/build/BUILD/libpod-9551f6bb379d4af56dfb63ddf0f3682e40a6694e/_build/src/github.com/containers/libpod/vendor/github.com/containers/buildah/imagebuildah/chroot_symlink.go:142 +0x39b
github.com/containers/libpod/vendor/github.com/containers/buildah/imagebuildah.resolveSymlinkTimeModified()
	/builddir/build/BUILD/libpod-9551f6bb379d4af56dfb63ddf0f3682e40a6694e/_build/src/github.com/containers/libpod/vendor/github.com/containers/buildah/imagebuildah/chroot_symlink.go:98 +0x20a
github.com/containers/libpod/vendor/github.com/containers/storage/pkg/reexec.Init(0x116bf00)
	/builddir/build/BUILD/libpod-9551f6bb379d4af56dfb63ddf0f3682e40a6694e/_build/src/github.com/containers/libpod/vendor/github.com/containers/storage/pkg/reexec/reexec.go:26 +0x76
main.main()
	/builddir/build/BUILD/libpod-9551f6bb379d4af56dfb63ddf0f3682e40a6694e/_build/src/github.com/containers/libpod/cmd/podman/main.go:52 +0x6b
: exit status 2
```

What's curious and perhaps interests you is that if I use `buildah bud` then I'm not hitting this issue - which as an end-user I find somewhat confusing.. but I guess podman embeds a different version of buildah than when I directly invoke buildah on the CLI?! FYI I'm on Fedora 29 and using podman and buildah from packages:

```
$ buildah version
Version:         1.5
Go Version:      go1.11.2
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:  
Git Commit:      
Built:           Thu Jan  1 01:00:00 1970
OS/Arch:         linux/amd64

$ podman version
Version:       0.12.1.2
Go Version:    go1.11.2
Git Commit:    "67ab7549b44484cc3f201d7bb2b58b922f8edc24"
Built:         Thu Dec 13 20:35:26 2018
OS/Arch:       linux/amd64
```

It would be cool if a fix for this made it into the Fedora package so that I can help (promote the use of)  `podman-docker` instead of requiring Docker!

**Comment by vorburger** [comment]
> We have builds of v1.0 in Bodhi - should be hitting updates-testing soon.

perhaps it would be worthwhile to document how a noob (me) can easily get podman from _Bodhi_ (whatever that is, sorry), e.g. on https://github.com/containers/libpod/blob/master/docs/tutorials/podman_tutorial.md#install-podman-on-fedora-from-rpm-repositories ?  If that even possible, perhaps this suggestion doesn't really make sense; I dunno - just playing devil's advocate as a user here.

>  so that I can help (promote the use of) podman-docker instead of requiring Docker!

FYI https://git.opendaylight.org/gerrit/#/c/79589/2/odlCNIPlugin/odlovs-cni/Readme.md

**Comment by mheon** [comment]
@vorburger Bodhi is the Fedora package build system - the builds are gated there to make sure they're good before being released into the main repos. It would be good to document, though, so we can get community assistance in testing new builds and getting them out into the repos.

**Comment by rhatdan** [comment]
Podman 1.0 should be available in fedora 1.0 updates-testing now.

dnf update podman --enablerepo updates-testing

podman 1.0 on f29 just got enough Karma to be pushed to release.  F28 needs one more.

@vorburger OpenDaylight page looks great, thanks.

**Comment by rhatdan** [comment]
It is available now.
